gateway-service-impl/src/main/java/org/hypertrace/gateway/service/EntityTypesProvider.java

```diff
@@ -0,0 +1,64 @@
+package org.hypertrace.gateway.service;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import lombok.NonNull;
+import org.hypertrace.core.grpcutils.context.ContextualKey;
+import org.hypertrace.gateway.service.common.util.EntityTypeServiceClient;
+import org.hypertrace.gateway.service.common.util.EntityTypeServiceV2Client;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class EntityTypesProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(EntityTypesProvider.class);
+  private static final int DEFAULT_CACHE_SIZE = 1024;
+  private static final int DEFAULT_EXPIRE_DURATION_MIN = 30; // 30 minutes
+
+  // TenantId to Entity types cache
```

Comment:
the tenant ID part should be opaque, that's why we're using a context key. Suggest just calling this an entityTypeCache